cornell-dti/office-hours
# Undo Action Bugs

### Inside This PR 


* Fixed issue where students would get assigned to themselves if they retracted their own question and clicked "undo"
* Fixed minor issue where a TA clicking either "no show" or "done" on an assigned question and undoing the action would set the student's question back to "unresolved" rather than "assigned"

### Test Plan 
* User is a student:
  * Check that the question's status is correctly set to "unresolved" after retracting and undoing one's own question. 
  * Ensure that the student cannot retract their question when their question is marked as "assigned"
* User is a TA/Prof:  
  * Check that the question's status is correctly set to "unresolved" after undoing "Assign to Me" or "I really don't know" 
  * Check that the question's status is correctly set to "assigned" after undoing "No Show" or "Done" 

### Notes 

* When updating a question by using either updateQuestion or updateQuestionByQuestionId, the trigger_before_update_question function (found in 0001_base_schema.sql) will cause the current user to be automatically set as the answerer_id of that question, and also update time_addressed to the current time.
* This is used in the case where a TA/Prof assigns a student's question to themselves by updating question status to "assigned", and is therefore set as the answerer_id.
* However, a student user can also set themselves as the answerer_id if they retract their own question for example, as the question status is updated from "unresolved" to "retracted". This is not normally an issue if the status is set correctly, since display logic is usually done by checking a question's status.
* The "assign to self" bug came from a student user retracting their own question and undoing that action. The question should have been set to "unresolved", but was instead set to "assigned". As they were the last user to update that question, they were automatically set as the answerer_id. This was fixed by correctly setting the question status to "unresolved" after undoing.
* Additionally, a student can retract and undo their question when they are currently assigned to a TA. Then, the question's status should actually be set to "assigned", but would run into the bug described above. Since it is unlikely that a student will retract their question when being helped, the "remove" button is hidden in that case.
* Lastly, there were some cases where a TA/Prof's use of the undo button caused a question's status to be set incorrectly, which has also been addressed.
### Checklist




- [ ] My changes requires a change to the documentation.
